homeworks/solutions/krasnor/3-smooth/src/halfedge.js

```diff
@@ -0,0 +1,526 @@
+   * constructing the Laplace matrix. Possible value could be:
+   * 'uniform', 'cotan'.
+   */
+  laplaceMatrix(weightType) {
+    // TODO: construct the Laplace matrix.
+
+    const n = this.vertices.length
+    let T = new Triplet(n, n)
+
+    for (const vert of this.vertices) {
+      const i = vert.idx
+      let sum = 1e-8 // Tikhonov regularization to get strict positive definite
+      vert.halfedges(h => {
+        let w = 0
+        switch (weightType) {
+          case 'uniform':
+            w = 1
+            break
+          case 'cotan':
+            w = (h.cotan() + h.twin.cotan())/2
+        }
+        sum += w
+        T.addEntry(w, i, h.twin.vertex.idx)
+      })
+      T.addEntry(-sum, i, i)
+    }
+    return SparseMatrix.fromTriplet(T)
+  }
+
+  massMatrix(weightType){
+    const n = this.vertices.length;
+    let T = new Triplet(n, n);
+
+    for (const vert of this.vertices) {
+      const i = vert.idx;
+
+      let w = 0;
+      switch (weightType) {
+        case 'uniform':
+          // count neighbors
+          vert.halfedges(() => {
+            w++
+          });
+          T.addEntry(w,i,i);
+          break;
+        case 'cotan':
+          // surrounding area
+          w = vert.voronoiCell();
+          // if I implement it according to the slides (only Area) the result becomes too smooth
+          T.addEntry(1+w,i,i); // only w is too small -> too big changes in mesh; 1/w too big -> no noticeable changes
+          break;
+      }
+    }
+    return SparseMatrix.fromTriplet(T);
+  }
+
+  /**
+   * smooth performs the Laplacian smoothing algorithm.
+   * @param {string} weightType indicates the type of the weight for
+   * constructing the Laplace matrix. Possible value could be: 'uniform',
+   * 'cotan'.
+   * @param {Number} timeStep the time step in Laplacian Smoothing algorithm
+   * @param {Number} smoothStep the smooth step in Laplacian Smoothing algorithm
+   */
+  smooth(weightType, timeStep, smoothStep) {
+    // TODO: implmeent the Laplacian smoothing algorithm.
+    //
+    // Hint:
+    //
+    //   1. Construct the Laplace matrix `L` for the given `weightType`
+    //   2. Solve linear equation: (I-tλL) f' = f using a Cholesky solver.
+    //   3. Update the position of mesh vertices based on the solution f'.
+    //
+    this.resetVertexArray();
+
+    console.log("smoothing: %s %f %f", weightType, timeStep, smoothStep)
+    for (let s_step = 0; s_step < smoothStep; s_step++) {
+      const n_verts = this.vertsOrig.length;
+
+      // 1.
+      let laplaceM = this.laplaceMatrix(weightType);
+      let massM = this.massMatrix(weightType);
+
+      // 2.
+      let f = massM.minus(laplaceM.timesReal(timeStep));
+
+      let b = DenseMatrix.zeros(n_verts, 3)
+      for (const v of this.vertices) {
+        b.set(v.position.x, v.idx, 0);
+        b.set(v.position.y, v.idx, 1);
+        b.set(v.position.z, v.idx, 2);
+      }
+      let result = f.chol().solvePositiveDefinite(massM.timesDense(b));
```

Comment:
As an additional note (for your further reading):

The smooth step and time step can be improved. In each smooth step, the Laplace matrix is recomputed.
This is essential because the positions are updated after each step, and the matrix should also be updated/recomputed.

There are research papers actually investigate what could happen if the Laplace matrix is only computed once (always use the original Laplace). See, for instance:

Kazhdan, Michael, Jake Solomon, and Mirela Ben Chen.
"Can mean curvature flow be modified to be non-singular?."
Computer Graphics Forum. 2012. https://arxiv.org/pdf/1203.6819.pdf